common/src/main/java/net/william278/husksync/database/MongoDbDatabase.java

```diff
@@ -0,0 +1,357 @@
+                        try {
+                            Document filter = new Document("uuid", existingUser.getUuid().toString());
+                            Document doc = mongoCollectionHelper.getCollection(usersTable).find(filter).first();
+
+                            Bson updates = Updates.set("uuid", user.getUuid().toString());
+                            mongoCollectionHelper.updateDocument(usersTable, doc, updates);
+                        } catch (MongoException e) {
+                            plugin.log(Level.SEVERE, "Failed to insert a user into the database", e);
+                        }
+                    }
+                },
+                () -> {
+                    // Insert new player data into the database
+                    try {
+                        Document doc = new Document("uuid", user.getUuid().toString()).append("username", user.getUsername());
+                        mongoCollectionHelper.insertDocument(usersTable, doc);
+                    } catch (MongoException e) {
+                        plugin.log(Level.SEVERE, "Failed to insert a user into the database", e);
+                    }
+                }
+        );
+    }
+
+    /**
+     * Get a player by their Minecraft account {@link UUID}
+     *
+     * @param uuid Minecraft account {@link UUID} of the {@link User} to get
+     * @return An optional with the {@link User} present if they exist
+     */
+    @Blocking
+    @Override
+    public Optional<User> getUser(@NotNull UUID uuid) {
+        Document filter = new Document("uuid", uuid);
+        Document doc = mongoCollectionHelper.getCollection(usersTable).find(filter).first();
+        if (doc != null) {
+            return Optional.of(new User(UUID.fromString(doc.getString("uuid")),
+                    doc.getString("username")));
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Get a user by their username (<i>case-insensitive</i>)
+     *
+     * @param username Username of the {@link User} to get (<i>case-insensitive</i>)
+     * @return An optional with the {@link User} present if they exist
+     */
+    @Blocking
+    @Override
+    public Optional<User> getUserByName(@NotNull String username) {
+        Document filter = new Document("username", username);
+        Document doc = mongoCollectionHelper.getCollection(usersTable).find(filter).first();
+        if (doc != null) {
+            return Optional.of(new User(UUID.fromString(doc.getString("uuid")),
+                    doc.getString("username")));
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Get the latest data snapshot for a user.
+     *
+     * @param user The user to get data for
+     * @return an optional containing the {@link DataSnapshot}, if it exists, or an empty optional if it does not
+     */
+    @Blocking
+    @Override
+    public Optional<DataSnapshot.Packed> getLatestSnapshot(@NotNull User user) {
```

Comment:
add NotNull annotation

Reply:
where do you want the notnull??? it already exists